lib/rmg/ObjectManager.cpp

```diff
@@ -0,0 +1,388 @@
+		rmg::Object rmgObject(*obj);
+		rmgObject.setTemplate(zone.getTerrainType());
+		bool guarded = addGuard(rmgObject, object.second, (obj->ID == Obj::MONOLITH_TWO_WAY));
+		
+		if(!placeAndConnectObject(zone.areaPossible(), rmgObject,
+								[this, &rmgObject](const int3 & tile)
+								{
+									float dist = rmgObject.getArea().distanceSqr(zone.getPos());
+									dist *= (dist > 12.f * 12.f) ? 10.f : 1.f; //tiles closer 12 are preferrable
+									dist = 1000000.f - dist; //some big number
+									return dist + map.getNearestObjectDistance(tile);
+								}, guarded, false, true))
+		{
+			logGlobal->error("Failed to fill zone %d due to lack of space", zone.getId());
+			return false;
+		}
+		
+		placeObject(rmgObject, guarded, true);
+		
+		for(const auto & nearby : nearbyObjects)
+		{
+			if(nearby.second != obj)
+				continue;
+			
+			rmg::Object rmgNearObject(*nearby.first);
+			rmg::Area possibleArea(rmgObject.instances().front()->getBlockedArea().getBorderOutside());
+			possibleArea.intersect(zone.areaPossible());
+			if(possibleArea.empty())
+			{
+				rmgNearObject.clear();
+				continue;
+			}
+			
+			rmgNearObject.setPosition(*RandomGeneratorUtil::nextItem(possibleArea.getTiles(), generator.rand));
+			placeObject(rmgNearObject, false, false);
+		}
+	}
+	
+	//create object on specific positions
+	//TODO: implement guards
+	for (const auto &obj : instantObjects)
+	{
+		rmg::Object rmgObject(*obj.first);
+		rmgObject.setPosition(obj.second);
+		placeObject(rmgObject, false, false);
+	}
+	
+	requiredObjects.clear();
+	closeObjects.clear();
+	nearbyObjects.clear();
+	instantObjects.clear();
+	
+	return true;
+}
+
+void ObjectManager::placeObject(rmg::Object & object, bool guarded, bool updateDistance)
+{
+	assert(!zone.areaUsed().overlap(object.getArea()));
+	
+	object.finalize(map);
+	zone.areaPossible().subtract(object.getArea());
+	bool keepVisitable = zone.freePaths().contains(object.getVisitablePosition());
+	zone.freePaths().subtract(object.getArea()); //just to avoid areas overlapping
+	if(keepVisitable)
+		zone.freePaths().add(object.getVisitablePosition());
+	zone.areaUsed().unite(object.getArea());
+	zone.areaUsed().erase(object.getVisitablePosition());
+	
+	if(guarded)
+	{
+		auto guardedArea = object.instances().back()->getAccessibleArea();
+		guardedArea.add(object.instances().back()->getVisitablePosition());
+		auto areaToBlock = object.getAccessibleArea(true);
+		areaToBlock.subtract(guardedArea);
+		zone.areaPossible().subtract(areaToBlock);
+		for(auto & i : areaToBlock.getTilesVector())
+			if(map.isOnMap(i) && map.isPossible(i))
+				map.setOccupied(i, ETileType::BLOCKED);
+	}
+	
+	if(updateDistance)
+		updateDistances(object);
+	
+	switch(object.instances().front()->object().ID)
+	{
+		case Obj::TOWN:
+		case Obj::RANDOM_TOWN:
+		case Obj::MONOLITH_TWO_WAY:
+		case Obj::MONOLITH_ONE_WAY_ENTRANCE:
+		case Obj::MONOLITH_ONE_WAY_EXIT:
+		case Obj::SUBTERRANEAN_GATE:
+		case Obj::SHIPYARD:
+		case Obj::MINE:
```

Comment:
Since when are mines connected with roads?

Reply:
Was not it your idea to connect them with invisible roads to town to ensure their reachability?

Reply:
Just for debug purposes, will be removed. Algorithms ensures accessibility for all objects

Reply:
What about option to actually place lower quality roads for connection between castles and mines?

Reply:
AFAIK, there are options to enable/disable different road types in hot rmg. Does anybody know what's the logic beyond this?